src/pycsou/opt/solver/lsqr.py

```diff
@@ -0,0 +1,311 @@
+        mst["cs2"] = -xp.ones_like(beta)
+        mst["rhobar"] = mst["alpha"] = alpha
+        mst["phibar"] = mst["rnorm"] = mst["normr1"] = mst["normr2"] = beta
+        mst["bnorm"] = bnorm
+        mst["test1"], mst["test2"] = xp.ones_like(beta), alpha / beta
+        mst["test3"] = mst["t1"] = mst["rtol"] = None
+
+        # Vectorize the function of Givens rotation
+        self._sym_ortho_func = xp.vectorize(self._sym_ortho)
+
+    @staticmethod
+    def _sym_ortho(a, b, xp):
+        """
+        Stable implementation of Givens rotation.
+        """
+        if b == 0:
+            return xp.sign(a), 0, xp.abs(a)
+        elif a == 0:
+            return 0, xp.sign(b), xp.abs(b)
+        elif xp.abs(b) > xp.abs(a):
+            tau = a / b
+            s = xp.sign(b) / xp.sqrt(1 + tau * tau)
+            c = s * tau
+            r = b / s
+        else:
+            tau = b / a
+            c = xp.sign(a) / xp.sqrt(1 + tau * tau)
+            s = c * tau
+            r = a / c
+        return c, s, r
+
+    def m_step(self):
+
+        mst = self._mstate  # shorthand
+
+        xp = pycu.get_array_module(mst["x"])
+
+        x, u, v, w = mst["x"], mst["u"], mst["v"], mst["w"]
+        alpha, rhobar, phibar = mst["alpha"], mst["rhobar"], mst["phibar"]
+        normA, ddnorm, xxnorm, bnorm = mst["normA"], mst["ddnorm"], mst["xxnorm"], mst["bnorm"]
+        res2, sn2, cs2, z = mst["res2"], mst["sn2"], mst["cs2"], mst["z"]
+
+        # Bidiagonalizaion to obtain next beta, u, alpha, v
+        u = self._A.apply(v) - alpha * u
+        beta = xp.linalg.norm(u, axis=-1, keepdims=True)
+
+        if (beta > 0).all():
+            u *= 1 / beta
+            normA = xp.sqrt(normA**2 + alpha**2 + beta**2 + self._dampsq)
+            v = self._A.T.apply(u) - beta * v
+            alpha = xp.linalg.norm(v, axis=-1, keepdims=True)
+            if (alpha > 0).all():
+                v = (1 / alpha) * v
+
+        # Plane rotation to eliminate the damping parameter
+        if self._damp > 0:
+            rhobar1 = xp.sqrt(rhobar**2 + self._dampsq)
+            cs1 = rhobar / rhobar1
+            sn1 = self._damp / rhobar1
+            psi = sn1 * phibar
+            phibar = cs1 * phibar
+        else:
+            rhobar1 = rhobar
+            psi = xp.zeros_like(beta)
+
+        # Plane rotation to eliminate the subdiagonal element of lower-bidiagonal matrix
+        cs, sn, rho = self._sym_ortho_func(rhobar1, beta, xp)
+
+        theta = sn * alpha
+        rhobar = -cs * alpha
+        phi = cs * phibar
+        phibar = sn * phibar
+        tau = sn * phi
+
+        # Update x and w
+        t1 = phi / rho
+        t2 = -theta / rho
+        dk = (1 / rho) * w
+        x = x + t1 * w
+        w = v + t2 * w
+
+        mst["x"], mst["u"], mst["v"], mst["w"] = x, u, v, w
+        mst["alpha"], mst["rhobar"], mst["phibar"] = alpha, rhobar, phibar
+
+        # Estimation of norms:
+        ddnorm = ddnorm + xp.linalg.norm(dk) ** 2
```

Comment:
which axis should the norm be applied? Dask will only work for 1D or 2D inputs when axis is not specified.